### DISCUSSION TITLE
Wrong center of mass for clipped body

### DISCUSSION THREAD

**Discussion by vzavec** [pyvista/pyvista#5524] [Q&A]
Hi there!

Just step into PYVISTA world and need help on my code example bellow.
From csv file I received XYZ data for ship hull. 
There is plane plz for making clip and slice situation.
I am in doubt about the clipped volume, but most of all I think that mass center of clipped body must have negative value.

May I ask for some suggestion on my code.

```python
import numpy as np
import pyvista as pv
import pandas as pd

# plane center
center = (160, 0, 20)

# plane dimension
i_sz=360
j_sz=80

direction = [0.0, 0.26723838, 0.96363045]
plz = pv.Plane(
        center=center,
        direction=direction,
        i_size=i_sz,
        j_size=j_sz,
    )

# ship hull section data from csv file
df = pd.read_csv("ladja_pyvista.csv", sep=';', header=None) 
df.columns = ['x', 'y', 'z']
np_points = np.c_[df['x'].ravel(), df['y'].ravel(), df['z'].ravel()]

cloud = pv.PolyData(np_points)
surf = cloud.delaunay_3d().clean(tolerance=1e-10)
surf.plot(show_grid=True, show_edges=False, color=[155, 255, 125], lighting=True, opacity=0.85)

print("Hull volume (m3) : ", surf.volume)
print("Mass center : ", surf.center_of_mass())

# contact area between plane plz and clipped body
slc = surf.slice(normal=direction, origin=center)
print("Slice area (m2) : ", slc.area)

pl = pv.Plotter()
actor = pl.show_grid(
    color='gray',
    location='outer',
    grid='back',
    ticks='outside',
    xtitle='Dolžina (m)',
    ytitle='Širina (m)',
    ztitle='Višina (m)',
    font_size=12,
)
actor = pl.add_mesh(plz, opacity=0.35, color='blue')         
actor = pl.add_mesh(surf, color='yellow', opacity=0.35)
actor = pl.add_mesh(slc, color='black', opacity=0.75)
actor = pl.add_axes()
pl.show()

# clip surface with plane plz
ccl = surf.clip_surface(plz, invert=True)
triangulated = ccl.delaunay_3d().extract_surface()
triangulated.plot(show_edges=True, color="green", lighting=True, opacity=0.25, show_grid=True, line_width=3) 

print("Volume - clipped (m3) : ", triangulated.volume)

# must have negative y value  ????
print("Mass center clipped : ", triangulated.center_of_mass())
```

**Comment by vzavec** [comment]
With the direction = [0. 0,  0.38268343,  0.92387953] the Mass center clipped is :  [168.43605222   7.09908215  14.71752402]. Y-axis must be negative! 

<img width="237" alt="situation" src="https://github.com/pyvista/pyvista/assets/20023991/7ca58bc7-5b4c-46de-a4f8-3e84294c10fa">

Forgot to include my data file: 
[ladja_pyvista.csv](https://github.com/pyvista/pyvista/files/14064365/ladja_pyvista.csv)

**Comment by MatthewFlamm** [comment]
The documentation in pyvista needs to be updated, but `center_of_mass` does not in fact yield the center of mass as you are expecting.  It follows the naming scheme of the vtk method `vtkCenterOfMass` which gives the center of mass of the _points_ of the mesh only. That is it does not use the cellular structure.  This would be useful for example if each point were an atom with a particular atomic weight.

Here is a minimally reproducible example that illustrates this, and gives my suggested way to calculate the center of mass using a 3D mesh.

```py
import pyvista as pv
import numpy as np

# generate a mesh that is heavily weighted with points towards the bottom in z direction
# true center of mass is at [0, 0, 0]
phi = np.linspace(0, 90, 5)
phi = np.append(phi[0:-1], np.linspace(90, 180, 100))
mesh = pv.SolidSphereGeneric(phi=phi)
print(mesh.extract_surface().center_of_mass())  # close to [0, 0, -0.3]

# use integrate data on the 3D space filling mesh to get better estimate of center of mass
mesh["points"] = mesh.points
int_mesh = mesh.integrate_data()
print(int_mesh["points"]/int_mesh["Volume"])  # close to [0, 0, 0] to about ~ 2-3 digits.

# make plot to illustrate bottom heavy mesh in terms of points
pv.OFF_SCREEN=True
pl = pv.Plotter()
pl.add_mesh(mesh.extract_surface(), show_edges=True, scalars="points", component=2)
pl.view_xz()
pl.show(screenshot="tmp.png")
```
![image](https://github.com/pyvista/pyvista/assets/39341281/34833829-25a6-4c2d-8b09-041e3f0179f3)

**Reply by vzavec** [reply]
I'm glad for the answer.
Cutting your sphere with my plane shows results as you explained. 

But still do not know how to apply this to my shipp hull which has clipped volume filled with water.
And mass center of homogeneous watter is for me hard to compare with heavily weighted points from clipped surface.

>>> Mass center of clipped sphere:  [-9.37768975e-02 -2.88977237e-18 -1.50000906e-01]

Can i hope for more detailed advice?
<img width="446" alt="slika" src="https://github.com/pyvista/pyvista/assets/20023991/9ae2f9b6-7912-4640-9ca7-f3575920f5c2">

**Reply by MatthewFlamm** [reply]
Did you try the `integrate_data` approach?

**Reply by vzavec** [reply]
Yes, I tried the integrate_data approach and center of mass for clipped body is OK! 
Thanks!

**Reply by vzavec** [reply]
Can someone please give me a hint on how to calculate the moment of inertia of the slice surface from Pyvista function:

slc = surf.slice(normal=dirz, origin=center)
slc.plot(color='red', line_width=1, opacity=0.45, show_grid=True, show_edges=True)

Slice area is here:

<img width="460" alt="slice_area" src="https://github.com/pyvista/pyvista/assets/20023991/fa8ab95a-19a9-405a-86a4-f2314c45919d">
